mesos-client/src/main/scala/com/mesosphere/mesos/client/MesosClient.scala

```diff
@@ -209,37 +225,32 @@ object MesosClient extends StrictLogging with StrictLoggingFlow {
       askTimeout: Timeout): Source[(HttpResponse, Session), NotUsed] =
     urls match {
       case Nil => throw new IOException(s"Failed to connect to Mesos: List of master urls exhausted.")
-      case url :: rest =>
-        logger.info(s"Connecting to Mesos master $url")
-        connectionSource(frameworkInfo, url, authorization).map { response =>
+      case baseUrl :: rest =>
+        // Convert the Java URL to an Akka Uri.
+        val baseUri = Uri.from(
+          scheme = baseUrl.getProtocol,
+          host = baseUrl.getHost,
+          port = if (baseUrl.getPort == -1) baseUrl.getDefaultPort else baseUrl.getPort)
+        logger.info(s"Connecting to Mesos master $baseUri")
+        val requestUri = baseUri.withPath(Path("/api/v1/scheduler"))
+        connectionSource(frameworkInfo, requestUri, authorization, maxRedirects).map { response =>
           response.status match {
             case StatusCodes.OK =>
-              logger.info(s"Connected successfully to $url");
+              logger.info(s"Connected successfully to $baseUri");
               val streamId = response.headers
                 .find(h => h.is(MesosStreamIdHeaderName.toLowerCase))
                 .getOrElse(throw new IllegalStateException(s"Missing MesosStreamId header in ${response.headers}"))
 
-              (response, Session(url, streamId.value(), authorization))
-            case StatusCodes.TemporaryRedirect =>
-              val leader = new URL(response.header[headers.Location].get.value())
-              logger.warn(s"New Mesos leader available at $leader")
-              // Update the context with the new leader's host and port and throw an exception that is handled in the
-              // next `recoverWith` stage.
-              response.discardEntityBytes()
-              throw MesosRedirectException(leader)
+              (response, Session(baseUri, streamId.value(), authorization))
             case _ =>
               response.discardEntityBytes()
               throw new IllegalArgumentException(s"Mesos server error: ${response.status}")
           }
         }.recoverWithRetries(
           1, {
-            case ex @ MesosRedirectException(leader) =>
-              if (maxRedirects > 0)
-                mesosHttpConnection(frameworkInfo, (leader :: rest).distinct, maxRedirects - 1, authorization)
-              else
-                throw new IOException("Failed to connect to Mesos: Too many redirects.", ex)
             case ex =>
-              logger.warn(s"Failed to connect to Mesos $url", ex)
+              // TODO: This retry only works on the initial connection. It does *not* work the the connection dies.
```

Comment:
@kaiwalyajoshi and @timcharper. When need to design this better. In HA mode I would expect that the client tries to connect to another master URL when the initial connection dies. Currently, this is not the case.

Reply:
Ping @kaiwalyajoshi and @timcharper

Reply:
Yeah, that was the intention of the code here. I'd prefer keep "following the leader" and recovering from errors separate concerns. Let's wrap this connection with another retrying stream.

Reply:
I suppose we'll get a new subscribed message letting us know that a reconnect occurred? And this logic will take care of re-sending that subscribe?

I'm inclined to want to have the re-connect in the event of failure logic outside of the redirection following logic, but I haven't really thought this fully through. What's your impression?

Reply:
I think the stream should die and we provide a recipe in the docs on how to restart the flow. Take a look at the [Mesos Plugin](https://github.com/jenkinsci/mesos-plugin/pull/394/files#diff-812b2ba3c16e3644eaa01ca458ee94a4R68). Since the restart also implies how to recreate the state USI cannot restart on its own. However, USI should try another master if the connection fails.